orca-keel/src/main/kotlin/com/netflix/spinnaker/orca/keel/task/PublishDeliveryConfigTask.kt

```diff
@@ -0,0 +1,156 @@
+ * to support GitOps flows.
+ */
+@Component
+class PublishDeliveryConfigTask
+constructor(
+  private val keelService: KeelService,
+  private val scmService: ScmService,
+  private val objectMapper: ObjectMapper
+) : RetryableTask {
+  private val log = LoggerFactory.getLogger(javaClass)
+
+  override fun execute(stage: Stage): TaskResult {
+    val context = objectMapper.convertValue<PublishDeliveryConfigContext>(stage.context)
+    val trigger = stage.execution.trigger
+    val user = trigger.user ?: "anonymous"
+
+    if (context.attempt > context.maxRetries!!) {
+      val error = "Maximum number of retries exceeded (${context.maxRetries})"
+      log.error("$error. Aborting.")
+      return TaskResult.builder(ExecutionStatus.TERMINAL).context(mapOf("error" to error)).build()
+    }
+
+    // if the pipeline has a git trigger, infer what context we can from the trigger
+    if (trigger is GitTrigger) {
+      // FIXME: GitTrigger props are non-null, but IIUC branch and hash would be mutually exclusive, so what to do here?
+      if (trigger.hash != null && context.ref == null) {
+        context.ref = trigger.hash
+        log.debug("Inferred context.ref from trigger: ${context.ref}")
+      } else if (trigger.branch != null && context.ref == null) {
+        context.ref = "refs/heads/${trigger.branch}"
+        log.debug("Inferred context.ref from trigger: ${context.ref}")
+      }
+      if (context.scmType == null) {
+        context.scmType = trigger.source
+        log.debug("Inferred context.scmType from trigger: ${context.scmType}")
+      }
+      if (context.project == null) {
+        context.project = trigger.project
+        log.debug("Inferred context.project from trigger: ${context.project}")
+      }
+      if (context.repository == null) {
+        context.repository = trigger.slug
+        log.debug("Inferred context.repository from trigger: ${context.repository}")
+      }
+    } else {
+      if (context.ref == null) {
+        context.ref = "refs/heads/master"
+      }
+      if (context.scmType == null || context.project == null || context.repository == null) {
+        throw IllegalArgumentException("scmType, project and repository are required fields in the stage if there's no git trigger.")
+      }
+    }
+
+    val manifestLocation = "${context.scmType}://${context.project}/${context.repository}/<manifestBaseDir>/${context.directory
+      ?: ""}/${context.manifest}@${context.ref}"
+
+    return try {
+      log.debug("Retrieving keel manifest at $manifestLocation")
+      val deliveryConfig = scmService.getDeliveryConfigManifest(
+        context.scmType, context.project, context.repository, context.directory, context.manifest, context.ref)
+
+      log.debug("Publishing manifest ${context.manifest} to keel on behalf of $user")
+      keelService.publishDeliveryConfig(deliveryConfig, user)
+
+      TaskResult.builder(ExecutionStatus.SUCCEEDED).context(emptyMap<String, Any?>()).build()
```

Comment:
This is really the meat of the task... It's fairly simple: pull the delivery config from igor, push to keel.